### PR TITLE
[pipeline] test pure pipeline process using llama

### LINKED ISSUE #4204: [shardformer] support pipeline for llama
add llama forward and policy

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
closed #4204 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR adds a pure pipeline test process using llama model

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [x] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/p2p.py [hunk 1]
```diff
@@ -64,7 +64,10 @@ def _broadcast_object_list(object_list: List[Any],
     my_rank = dist.get_rank()
     # Serialize object_list elements to tensors on src rank.
     if my_rank == src:
-        tensor_list, size_list = zip(*[c10d._object_to_tensor(obj) for obj in object_list])
+        if torch.__version__ >= "1.13.0":
```
Comment (ver217) [inline colossalai/pipeline/p2p.py:67]: image
FYI

File: tests/test_shardformer/test_model/test_pure_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    return_loss: bool = True,
+    return_outputs: bool = False,
+    schedule: OneForwardOneBackwardSchedule = None,
+) -> dict:
+    # return loss or outputs if needed
+    outputs = schedule.forward_backward_step(model, optimizer, data_iter, criterion, return_loss, return_outputs)
+    return outputs
+
+
+class data_iter():
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_pure_pipeline.py:97]: This naming is bad

File: tests/test_shardformer/test_model/test_pure_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        #dataloader=prepare_dataloader(dataset=dataset['train'],batch_size=4)
+        schedule = OneForwardOneBackwardSchedule(num_microbatches, stage_manager)
+        shard_config = ShardConfig(enable_fused_normalization=enable_fused_normalization,
+                                   enable_tensor_parallelism=enable_tensor_parallelism,
+                                   pipeline_stage_manager=stage_manager)
+        pipelined_model = PipelinedModel(org_model, shard_config, stage_manager)
+        pp_optimizer = PipelineOptimizer(optimizer, pipelined_model)
+        data_it = iter(data_iter())
+        results = execute_pipeline(data_it, pipelined_model, loss, pp_optimizer, schedule=schedule)
+        assert results['loss'] is not None
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_pure_pipeline.py:144]: You should check whether the loss is correct

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/pipeline/p2p.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/pipeline/p2p.py:67]: done

File: tests/test_shardformer/test_model/test_pure_pipeline.py (hunk 2 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_pure_pipeline.py:97]: modified

File: tests/test_shardformer/test_model/test_pure_pipeline.py (hunk 3 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_pure_pipeline.py]: left a check here and find some bug

File: tests/test_shardformer/test_model/test_pure_pipeline.py [hunk 4]
```diff
@@ -125,22 +127,35 @@ def run_llama_test(enable_fused_normalization, enable_tensor_parallelism, use_la
...
                                    pipeline_stage_manager=stage_manager)
         pipelined_model = PipelinedModel(org_model, shard_config, stage_manager)
         pp_optimizer = PipelineOptimizer(optimizer, pipelined_model)
-        data_it = iter(data_iter())
-        results = execute_pipeline(data_it, pipelined_model, loss, pp_optimizer, schedule=schedule)
+        results = execute_pipeline(data_iter, pipelined_model, loss, pp_optimizer, schedule=schedule)
+
         if stage_manager.is_last_stage():
-            assert results['loss'] is not None
+            assert results['loss'] == -org_loss
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_pure_pipeline.py:155]: What does this mean?

Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_pure_pipeline.py]: done

**Review by ver217 — APPROVED** [review]

**Comment by CjhHa1** [comment]
I left the check unsuccessful and the modified test should be tested after merge

**Pull request merged**